README.md
# Lavaboom API

[![Code Climate](https://codeclimate.com/github/lavab/api/badges/gpa.svg)](https://codeclimate.com/github/lavab/api)

<img src="https://mail.lavaboom.com/img/Lavaboom-logo.svg" align="right" width="200px" />

Lavaboom's main API written in Golang.

Contains the core Lavaboom's functionality. Currently a monolith, which later
will be split and transformed into a set of microservices. 

## Requirements

 - Redis
 - RethinkDB
 - NSQ

## Installation

### Inside a Docker container

*This image will be soon uploaded to Docker Hub.*

```bash
git clone https://github.com/lavab/api.git
cd api
docker build -t "lavab/api" .
docker run \
	-p 127.0.0.1:5000:5000 \
	--name api \
	lavab/api \
	-redis_address=172.8.0.1:6379 \
	-lookupd_address=172.8.0.1:4161 \
	-nsqd_address=172.8.0.1:4150 \
	-rethinkdb_address=172.8.0.1:28015 \
	-api_host=api.lavaboom.com \
	-email_domain=lavaboom.com
```

### Directly running the parts

```bash
go get github.com/lavab/api

api \
	-redis_address=172.8.0.1:6379 \
	-lookupd_address=172.8.0.1:4161 \
	-nsqd_address=172.8.0.1:4150 \
	-rethinkdb_address=172.8.0.1:28015 \
	-api_host=api.lavaboom.com \
	-email_domain=lavaboom.com
```

## Passing configuration

You can use either commandline flags:
```
{ api } master » ./api -help
Usage of api:
  -api_version="v0": Shown API version
  -bind=":5000": Network address used to bind
  -config="": config file to load
  -email_domain="lavaboom.io": Domain of the default email service
  -etcd_address="": etcd peer addresses split by commas
  -etcd_ca_file="": etcd path to server cert's ca
  -etcd_cert_file="": etcd path to client cert file
  -etcd_key_file="": etcd path to client key file
  -etcd_path="settings/": Path of the keys
  -force_colors=false: Force colored prompt?
  -log="text": Log formatter type. Either "json" or "text"
  -lookupd_address="127.0.0.1:4160": Address of the lookupd server
  -nsqd_address="127.0.0.1:4150": Address of the nsqd server
  -redis_address="127.0.0.1:6379": Address of the redis server
  -redis_db=0: Index of redis database to use
  -redis_password="": Password of the redis server
  -rethinkdb_address="127.0.0.1:28015": Address of the RethinkDB database
  -rethinkdb_db="dev": Database name on the RethinkDB server
  -rethinkdb_key="": Authentication key of the RethinkDB database
  -session_duration=72: Session duration expressed in hours
  -slack_channel="#notif-api-logs": channel to which Slack bot will send messages
  -slack_icon=":ghost:": emoji icon of the Slack bot
  -slack_level="warning": minimal level required to have messages sent to slack
  -slack_url="": URL of the Slack Incoming webhook
  -slack_username="API": username of the Slack bot
  -yubicloud_id="": YubiCloud API id
  -yubicloud_key="": YubiCloud API key

```

Or environment variables:
```
{ api } master » BIND=127.0.0.1:6000 VERSION=v1 ./api
```

Or configuration files:
```
{ api } master » cat api.conf
# lines beggining with a "#" character are treated as comments
bind :5000
force_colors false
log text

rethinkdb_db dev
# configuration values can be empty
rethinkdb_key
# Keys and values can be also seperated by the "=" character
rethinkdb_url=localhost:28015

session_duration=72
version=v0
{ api } master » ./api -config api.conf
```

## License

This project is licensed under the MIT license. Check `license` for more
information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]
### Changed
  - Removed .drone and .travis

## [2.0.2] - 2015-05-19
### Added
 - Added a check whether an address mapping is used in the username
   reservation.
 - SockJS API client for headless serverside client development.
 - Onboarding emails that introduce users to the service.
 - Multiple identity support (also known as email aliases).

### Changed
 - Moved from traditional `go get`-based flow to dependency vendoring
   using [godep](https://github.com/tools/godep).
 - Disabled most of the log output to make it easier to analyze.
 - Matching for 10k most used passwords replaced with a bloom filter
   containing 17.5m leaked passwords from various hacks.

### Fixed
 - Cursor leakage all over the `db` package.
 - thread.update changing date_modified field of the model, which
   resulted in invalid ordering of the emails in the web client.
   Emails in "spam" being shown as unread on the sidebar (new label
   fetching query).
 - Incorrect difference checker in thread.update.

## [2.0.1] - 2015-04-15
### Added
 - Address mapping table for account's name-to-id lookups.
 - Username length check during registration.

### Changed
 - New index creation code (multiple compound and multi indexes).

### Fixed
 - Lack of Message-ID header causing Lavaboom emails to be flagged as
   spam.

## 2.0.0 - 2015-04-02
### Added
 - Initial release of Lavaboom API 2.0

[unreleased]: https://github.com/lavab/api/compare/2.0.2...HEAD
[2.0.2]: https://github.com/lavab/api/compare/2.0.2...2.0.1
[2.0.1]: https://github.com/lavab/api/compare/2.0.1...0.2.0

main.go
package main

import (
	"net"
	"net/http"
	"os"

	"github.com/Sirupsen/logrus"

	"github.com/lavab/flag"
	"github.com/zenazn/goji/graceful"

	"github.com/lavab/api/env"
	"github.com/lavab/api/setup"
)

// TODO: "Middleware that implements a few quick security wins"
// 		 https://github.com/unrolled/secure

var (
	// Enable namsral/flag functionality
	configFlag = flag.String("config", "", "config file to load")
	// General flags
	bindAddress      = flag.String("bind", ":5000", "Network address used to bind")
	apiVersion       = flag.String("api_version", "v0", "Shown API version")
	logFormatterType = flag.String("log", "text", "Log formatter type. Either \"json\" or \"text\"")
	forceColors      = flag.Bool("force_colors", false, "Force colored prompt?")
	emailDomain      = flag.String("email_domain", "lavaboom.io", "Domain of the default email service")
	// Registration settings
	sessionDuration = flag.Int("session_duration", 72, "Session duration expressed in hours")
	// Cache-related flags
	redisAddress = flag.String("redis_address", func() string {
		address := os.Getenv("REDIS_PORT_6379_TCP_ADDR")
		if address == "" {
			address = "127.0.0.1"
		}
		return address + ":6379"
	}(), "Address of the redis server")
	redisDatabase = flag.Int("redis_db", 0, "Index of redis database to use")
	redisPassword = flag.String("redis_password", "", "Password of the redis server")
	// Database-related flags
	rethinkdbAddress = flag.String("rethinkdb_address", func() string {
		address := os.Getenv("RETHINKDB_PORT_28015_TCP_ADDR")
		if address == "" {
			address = "127.0.0.1"
		}
		return address + ":28015"
	}(), "Address of the RethinkDB database")
	rethinkdbKey      = flag.String("rethinkdb_key", os.Getenv("RETHINKDB_AUTHKEY"), "Authentication key of the RethinkDB database")
	rethinkdbDatabase = flag.String("rethinkdb_db", func() string {
		database := os.Getenv("RETHINKDB_DB")
		if database == "" {
			database = "dev"
		}
		return database
	}(), "Database name on the RethinkDB server")
	// nsq and lookupd addresses
	nsqdAddress = flag.String("nsqd_address", func() string {
		address := os.Getenv("NSQD_PORT_4150_TCP_ADDR")
		if address == "" {
			address = "127.0.0.1"
		}
		return address + ":4150"
	}(), "Address of the nsqd server")
	lookupdAddress = flag.String("lookupd_address", func() string {
		address := os.Getenv("NSQLOOKUPD_PORT_4160_TCP_ADDR")
		if address == "" {
			address = "127.0.0.1"
		}
		return address + ":4160"
	}(), "Address of the lookupd server")
	// YubiCloud params
	yubiCloudID  = flag.String("yubicloud_id", "", "YubiCloud API id")
	yubiCloudKey = flag.String("yubicloud_key", "", "YubiCloud API key")
	// etcd
	etcdAddress  = flag.String("etcd_address", "", "etcd peer addresses split by commas")
	etcdCAFile   = flag.String("etcd_ca_file", "", "etcd path to server cert's ca")
	etcdCertFile = flag.String("etcd_cert_file", "", "etcd path to client cert file")
	etcdKeyFile  = flag.String("etcd_key_file", "", "etcd path to client key file")
	etcdPath     = flag.String("etcd_path", "settings/", "Path of the keys")
	// slack
	slackURL      = flag.String("slack_url", "", "URL of the Slack Incoming webhook")
	slackLevels   = flag.String("slack_level", "warning", "minimal level required to have messages sent to slack")
	slackChannel  = flag.String("slack_channel", "#notif-api-logs", "channel to which Slack bot will send messages")
	slackIcon     = flag.String("slack_icon", ":ghost:", "emoji icon of the Slack bot")
	slackUsername = flag.String("slack_username", "API", "username of the Slack bot")
	// Password bloom filter path
	bloomFilter = flag.String("bloom_filter", "bloom.db", "Bloom filter containing passwords")
	bloomCount  = flag.Uint("bloom_count", 14522336, "Estimated count of passwords in the bloom filter")
	// raven dsn
	ravenDSN = flag.String("raven_dsn", "", "DSN of the Raven connection")
)

func main() {
	// Parse the flags
	flag.Parse()

	// Put config into the environment package
	env.Config = &env.Flags{
		BindAddress:      *bindAddress,
		APIVersion:       *apiVersion,
		LogFormatterType: *logFormatterType,
		ForceColors:      *forceColors,
		EmailDomain:      *emailDomain,

		SessionDuration: *sessionDuration,

		RedisAddress:  *redisAddress,
		RedisDatabase: *redisDatabase,
		RedisPassword: *redisPassword,

		RethinkDBAddress:  *rethinkdbAddress,
		RethinkDBKey:      *rethinkdbKey,
		RethinkDBDatabase: *rethinkdbDatabase,

		NSQdAddress:    *nsqdAddress,
		LookupdAddress: *lookupdAddress,

		YubiCloudID:  *yubiCloudID,
		YubiCloudKey: *yubiCloudKey,

		SlackURL:      *slackURL,
		SlackLevels:   *slackLevels,
		SlackChannel:  *slackChannel,
		SlackIcon:     *slackIcon,
		SlackUsername: *slackUsername,

		BloomFilter: *bloomFilter,
		BloomCount:  *bloomCount,

		RavenDSN: *ravenDSN,
	}

	// Generate a mux
	mux := setup.PrepareMux(env.Config)

	// Make the mux handle every request
	http.Handle("/", mux)

	// Log that we're starting the server
	env.Log.WithFields(logrus.Fields{
		"address": env.Config.BindAddress,
	}).Info("Starting the HTTP server")

	// Initialize the goroutine listening to signals passed to the app
	graceful.HandleSignals()

	// Pre-graceful shutdown event
	graceful.PreHook(func() {
		env.Log.Info("Received a singnal, stopping the application")
	})

	// Post-shutdown event
	graceful.PostHook(func() {
		env.Log.Info("Stopped the application")
	})

	// Listen to the passed address
	listener, err := net.Listen("tcp", env.Config.BindAddress)
	if err != nil {
		env.Log.WithFields(logrus.Fields{
			"error":   err,
			"address": *bindAddress,
		}).Fatal("Cannot set up a TCP listener")
	}

	// Start the listening
	err = graceful.Serve(listener, http.DefaultServeMux)
	if err != nil {
		// Don't use .Fatal! We need the code to shut down properly.
		env.Log.Error(err)
	}

	// If code reaches this place, it means that it was forcefully closed.

	// Wait until open connections close.
	graceful.Wait()
}
